source/opt/aggressive_dead_code_elim_pass.cpp

```diff
@@ -178,24 +337,31 @@ bool AggressiveDCEPass::AggressiveDCE(ir::Function* func) {
   // If privates are not like local, add their stores to worklist
   if (!private_like_local_)
     for (auto& ps : private_stores_)
-      worklist_.push(ps);
+      AddToWorklist(ps);
   // Add OpGroupDecorates to worklist because they are a pain to remove
   // ids from.
   // TODO(greg-lunarg): Handle dead ids in OpGroupDecorate
   for (auto& ai : module_->annotations()) {
     if (ai.opcode() == SpvOpGroupDecorate)
-      worklist_.push(&ai);
+      AddToWorklist(&ai);
   }
   // Perform closure on live instruction set. 
   while (!worklist_.empty()) {
     ir::Instruction* liveInst = worklist_.front();
-    live_insts_.insert(liveInst);
     // Add all operand instructions if not already live
     liveInst->ForEachInId([this](const uint32_t* iid) {
       ir::Instruction* inInst = def_use_mgr_->GetDef(*iid);
-      if (live_insts_.find(inInst) == live_insts_.end())
-        worklist_.push(inInst);
+      if (!IsLive(inInst))
+        AddToWorklist(inInst);
     });
+    // If in a structured if construct, add the controlling conditional branch
+    // and its merge
+    ir::BasicBlock* blk = inst2block_[liveInst];
```

Comment:
I'm surprised this didn't walk up the stack of nested structured-ifs and mark all those ancestors live too.

What happens if there is a doubly-nested structured if,  and there is a live instruction inside the inner-if.  It seem the inner if branch (and merge) are marked live, but the outer one will still remain marked dead, and we'll accidentally delete the entire thing.
E.g.
```
  if (true) { // outer
     if (true) { // inner
        foo = 1; // something live.   the inst2block_ -> block2branch_ mapping will point to inner, but not outer
     }
  }
```
But then later the outer will be seen as dead and the whole thing swept away?

Reply:
Thinking about this more, this is handled because then the branch instruction at inner goes onto the work list and then we'll naturally end up walking up the stack.

Wow, this deserves a comment and a test case.